Adding method in GazeManager to send a message to object that is bein…

dbouianov opened the pull request:
…g focused on, which then can be added as a response in KeywordManager.

Here's how this is used from KeywordManager: http://i.imgur.com/PAGXMTu.png

The idea is to not to add a dependency on GazeManager inside KeywordManager.

Comments are welcome.

msftclas commented:
Hi __@dbouianov__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. Real humans will now evaluate your PR.
    </span>

TTYL, MSBOT;

NeerajW commented on Assets/HoloToolkit/Input/Scripts/GazeManager.cs:
```diff
@@ -129,5 +129,17 @@ private void UpdateStabilizationPlane()
+        /// Sends message to the object currently focused on.
+        /// </summary>
+        /// <param name="message">Message to send</param>
+        public void SendMessageToFocusedObject(string message)
```
Can you not do this already with OnGazeEnter and OnGazeExit?
I don't see too much benefit of this in the generic class. You could write your own MessageGazeManager since this is a partial class and add this function to your class.

dbouianov replied:
The problem is that the object field in Responses array is populated only statically.

How would OnGazeEnter and OnGazeExit be able to help? They only give you events that the object starts or ends to be gazed on, but these methods don't give you an ability to send custom messages to the focused object.

This is meant to be used from KeywordManager to send messages to focused object: http://i.imgur.com/PAGXMTu.png

jwittner replied:
An alternate solution would be a seperate small class which you register with the KeywordManager that uses the HitInfo available on the GazeManager to find the current gaze intersected object.

NeerajW replied:
Recommed making private GameObject focusedGameObject to a public read only property. Handle custom messages in your class.

dbouianov replied:
GazeManager already contains logic for detecting focusedObject, why not use data which is readily available? Besides, this is an elegant way to avoid leaking the GazeManager abstraction by not exposing FocusedObject outside of GazeManager.

NeerajW replied:
If you make the focusedObject public:
1. Any other class can also use it for hands / gestures or voice. They can do whatever they choose to do with it. 
2. They also don't need to depend on SendMessage VS SendMessageUpwards or other variants.
3. They could use it a more polymorphic way instead of this message way.

dbouianov replied:
@NeerajW Good point, I mentioned this in the forum thread. Exposing FocusedObject would create an additional dependency potentially used everywhere outside of GazeManager. This is probably fine, but PR intent was to minimize side effects and prevent the abstraction leaking outside of GazeManager.

jwittner replied:
A thought - maybe not as bad of a leak, but doesn't SendMessageToFocusedObject leak the abstraction similarly? It requires that GazeManager know the focused object and that the object it tracks is capable of receiving Unity style messages. This feels like the same contract as exposing the data through a readonly property.

dbouianov replied:
@jwittner Subtle but valid since GazeManager uses SendMessage only as a private implementation detail.

jwittner replied:
But to implement the functionality on the focused object you'd have to know that SendMessage was used correct?

dbouianov replied:
Not necessarily, you can add a public method and invoke it directly if you wanted to.

jwittner replied:
I meant your specific use case here - it relies on knowing that SendMessage is used.

dbouianov replied:
Oh, yes, this relies on SendMessage being used to enable dynamically created objects to receive recognized keywords without creating additional infrastructure for registering strongly-typed callbacks at runtime.

keveleigh commented:
The way we do this in Holograms 212 Voice is to add OnGazeEnter/Exit methods on the object, then turn on and off the KeywordManager appropriately. This way, we aren't filling our recognizer with commands that might not always be valid if, in this case, you arent gazing at anything.

keveleigh commented:
Actually, now that I keep thinking about this, this might be a really useful feature. My example above breaks if you want to send the same voice command to a variety of different objects, since you can't register the same command on more than one of them. This directs the result of the keyword recognizer to a specific object based on gaze instead.

dbouianov commented:
@keveleigh Does this mean each separate object needs to have a KeywordManager component attached to it? What if those objects are created dynamically, and there are 100 of them? Does that mean that there are 100 KeywordRecognizers being created?

dbouianov commented:
@keveleigh I think this would be efficient to create a single KeywordManager and let objects being gazed on decide whether they handle that keyword or not. This would prevent people from creating multiple KeywordRecognizers running simultaneously.

dbouianov commented:
With regards to filling the recognizer with invalid commands, it's basically up to the user in either of these scenarios.

NeerajW commented:
@dbouianov would it better for you to instead make private GameObject focusedObject a public get only property instead? That way you can handle custom send messages in your classes?

jwittner commented:
@NeerajW Exactly what I was going to say - it'd make lots of things easier! Otherwise I suspect we'll find more things like this being added to GazeManager.

dbouianov commented:
@NeerajW @jwittner Good points. I guess it would depending on how you see "better" and whether you're ready to expose GazeManager.FocusedObject and let other folks take the dependency. Agreed on leaking abstraction with SendMessageToFocusedObject is also letting clients know about FocusedObject. A (probably theoretical) concern with exposing FocusedObject is relying on Unity being single-threaded when setting FocusedObject in GazeManager.Update loop. We'll also see a lot of boilerplate all around: if (GazeManager.Instance.FocusedObject != null) GazeManager.Instance.FocusedObject.SendMessage (....), so I would probably keep both the public FocusedObject and SendMessageToFocusedObject as a convenience method that would avoid lots of boilerplate. Thoughts?

jwittner commented:
@dbouianov I think developers can keep the boilerplate centralized by creating simple behaviors like you'll have to for the SendMessage functionality wrapper in a "post exposed focus object" world. I'm not concerned about the single-threadedness of Update every changing and people can currently run code on background threads now so they'll have to be safe now anyways. 

I'd vote to still move SendMessageToFocused object out of GazeManager because the boilerplate can easily be encapsulated in an external object and we won't feel compelled to cover every type of useful interaction with game objects inside GazeManager.

dbouianov commented:
Updated PR, please review.

The intent is to add keyword recognition to multiple objects by simply sending messages to the currently focused object and letting the object decide whether to handle the event. Another intent is to minimize the amount of code the user has to write, so they can simply add a response in KeywordManager to make this work.

Please see this image for how this is intended to be used: http://i.imgur.com/aR0TdIb.png

This revision adds a separate class called FocusedObjectMessageSender, which I don't think is as elegant as simply adding a method to GazeManager.

Here's the previous way to enable this: http://i.imgur.com/PAGXMTu.png

I would lean in the direction of exposing both a public property GazeManager.FocusedObject and a convenience method GazeManager.SendMessageToFocusedObject which would avoid a lot of boilerplate in client code.

jwittner commented:
Looks like maybe you forgot to delete SendMessageToFocusedObject on the GazeManager?

NeerajW commented on Assets/HoloToolkit/Input/Scripts/FocusedObjectMessageSender.cs:
```diff
@@ -0,0 +1,17 @@
+﻿using UnityEngine;
+using HoloToolkit.Unity;
+
+public class FocusedObjectMessageSender : MonoBehaviour
```
Sorry to harp on this.
Feels like an overkill to have another script attached onto my game object to do something as simple as GazeManager.Instance.FocusedGameObject.SendMessage("FooBar");

I vote not having this script at all. And your change simply make the change to make focusedGameObject public.

@jwittner and @keveleigh please weigh in on your preference.

jwittner replied:
@NeerajW I think FocusedObjectMessageSender is great for the application level, but agree that it's overkill for the toolkit given it's simplicity.

dbouianov replied:
@NeerajW @jwittner @keveleigh So what's the consensus? Do we want add a separate class or add a convenience method to GazeManager?

jwittner replied:
My vote: Only the change to expose FocusedObject added to the HoloToolkit-Unity.
My suggestion: Use the class you wrote for your application.

=)

dbouianov replied:
This would defeat the purpose of helping everyone write less code by adding a simple way to add keyword recognition to dynamically created objects, because then everyone will need to add a variant of this to their code. Most interesting applications are not static, so we'll see a lot of this.

dbouianov commented:
Actually, it makes sense to decouple logic for sending messages in a separate class, since there are other potential policies for target object selection. Some examples: lazy focus with selection hysteresis, permanent focus with keyword to set the focused object. PR update coming up.

dbouianov commented:
Update PR, please review.

- Removed SendMessageToFocusedObject from GazeManager
- Added SelectedObjectMessageSender class
- Added FocusedObjectKeywordManager prefab
- Added SelectedObjectKeywordManager prefab

Rationale: users will want to create custom policies for selecting objects to send messages to that could be based on different mechanics: e.g. simple gaze, gaze hysteresis, selection voice commands.

jwittner commented:
I think with multiple examples this change gains a value in setting paradigm precedence. I think this is a good one to set for solving the multiple objects / single command voice control pattern, and by having it we avoid needing to have the discussion we just had again with other examples. If you can add a test scene showcasing both of these I'd be down to move it all into the HoloToolkit.

Thoughts everyone?

jwittner commented:
Here's a link to the general flow for contributions: https://github.com/Microsoft/HoloToolkit-Unity/blob/master/CONTRIBUTING.md

It outlines a few steps left in order to get this in. If you've got any questions, I'm sure the community has the answers, =)

dbouianov commented:
@jwittner Thanks for the contribution guide, I'll take a look. I'll also create a test scene to illustrate different object selection policies.

jwittner commented:
Hey @dbouianov just an FYI that I pulled your changes down and it looks like maybe you renamed the methods on SelectedObject, which broke your prefabs references. I'm sure you'd of caught it when building the test scene, but wanted to inform you anyhow. 

![image](https://cloud.githubusercontent.com/assets/12958442/16325346/3096f50a-3972-11e6-8942-7f9bc13bd5a1.png)

dbouianov commented:
Thanks, good catch, test scenes coming up.

NeerajW commented on ProjectSettings/EditorBuildSettings.asset:
```diff
@@ -4,4 +4,8 @@
   m_ObjectHideFlags: 0
   serializedVersion: 2
-  m_Scenes: []
+  m_Scenes:
```
@jwittner @dbouianov looks like a few things got missed in code review?
Please update this file to it's original settings.

NeerajW replied:
Also these files should not have been checked in.
Assets/HoloToolkit/Input/Models/Cursor/cursor_hand_ready.fbm/templates.meta  
Assets/HoloToolkit/Input/Models/Cursor/cursor_ready.fbm/templates.meta

NeerajW commented:
@jwittner @dbouianov looks like this got merged prematurely.

1. Additional files got checked in which should not have been updated. Added comment in the files.
2. The ReadMe.md was not updated.

NeerajW commented on Assets/HoloToolkit/Input/Tests/Prefabs/FocusedObjectSphere.prefab:
```diff
@@ -0,0 +1,117 @@
+%YAML 1.1
+%TAG !u! tag:unity3d.com,2011:
```
Why were so many prefabs added? How do they differ? Do we really need these?

jwittner replied:
Agreed, a lot of these seem like they should just be elements of the scene and don't require prefabs.

dbouianov replied:
OK, I'll remove unnecessary prefabs, just keep the examples.

jwittner commented:
I don't believe this was merged, just closed right?

dbouianov commented:
Yes, #69 was only closed following the issue/fix guidelines on the contribution guide, with #72 replacing the original commits.

> On Jun 24, 2016, at 1:16 PM, Josh Wittner <notifications@github.com> wrote:
> 
> I don't believe this was merged, just closed right?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Microsoft/HoloToolkit-Unity/pull/69#issuecomment-228451278>, or mute the thread <https://github.com/notifications/unsubscribe/ABYah9ICb80UueTnSUqg0-A2xN5xBVVsks5qPDsDgaJpZM4I8QBU>.
>

NeerajW commented:
Yeah I realized that later. Sorry about the panic :)
